### ISSUE TITLE
Channel shared between users

### ISSUE THREAD

**Issue by bristow** [Chocobozzz/PeerTube#699]
We are an french public administration, and it will very useful if we can have a videos channels shared between few users of our instances.

I don't think it's possible right now...

Thanks :+1:

**Comment by therahedwig** [comment]
Yes, this would be super helpful! For Krita it would mean we could have multiple people worry about the videos that go into a Krita channel without having to juggle accounts like we have to do for some social media, or the opposite, having to sent around heavy video files while the only thing it needs to do is to be sent to the peertube instance.

**Comment by BigMayo** [comment]
This would be very useful for the way we are using peertube in a community.
We have multiple users that would want to contribute different channels based on interests and groups/tasks.

**Comment by alcalyn** [comment]
It would be useful, and am wondering how to handle permissions between accounts and a channel ?

For now I'm thinking about a permissions system where we can manage users permissions:
    - add, edit, remove a video
    - edit or remove channel
    - add or remove accounts to the channel, and manage permissions

Or maybe there is a simpler system ?

Also it will modify the channel page, we could display all the account that are managing the channel

**Comment by mpfaff** [comment]
Are there any plans for this? I'd really appreciate a fine grained permission system in PeerTube, so that some users can only view a channel, while others can upload to it.

**Comment by 1000i100** [comment]
My use-case : 
I manage video contents for some clients, so i need acces and they also do to be as autonomous as they can/want.

I don't need fine permission managment, sharing admin acces to a channel is enouth for my use-case.

**Comment by k09i71** [comment]
I could see this issue going hand in hand with this other issue https://github.com/Chocobozzz/PeerTube/issues/2841

**Comment by test2a** [comment]
facebook pages have had fine grained permissions for a long time. 
https://www.facebook.com/help/289207354498410/
Having some control over permissions would be very helpful for this feature. suppose i want to add someone who can help moderate comments or edit tags and such but not upload videos on the channel or i want to give someone else permission to upload videos to the channel

**Comment by tassoman** [comment]
I think, to make the feature simpler, in first instance would be useful if the channel creator can add/remove users himself.
Users gains the ability to add/edit/remove all the videos in the channel.
I know for sure another bunch of public administrations interested on this feature, joined to LDAP/SAML2 plugin.

The need is to avoid the users being be the publicly owners of the video, the video should be linked to the channell itself. 
Or the owner should be the channel creator.

**Comment by test2a** [comment]
@tassoman that sounds reasonable enough to start with. as a channel "creator", you allow other users to work on videos. fine grained permissions can be later added which should not be much difficult. as #2841 issue exists, that would be helpful for teams in the long run. For federation, we could keep the creator as the "channel owner" and that would be distributed across activitypub but internally we could have many users,non public facing or public.

**Comment by kontrollanten** [comment]
I've the same use case as @1000i100 ; we're some users that want the same permissions to upload content to a channel. Our workaround right now is to have a Youtube channel that we're uploading to, then we've a cron job that runs PeerTube CLI to import the Youtube video. Pretty ugly, but it does the work.

What needs to be done to add this feature? Are there any big changes that needs to be done in the already existing data structures or functionality?

**Comment by stevenroose** [comment]
Is there any indication this is on the agenda? :) I'd also like to have this feature.

**Comment by rigelk** [comment]
It's not planned anytime soon

**Comment by svenzimmermann** [comment]
Would also be necessary for our use cases. We should have different uploaders in one channel. Also other users in the peer group should be able to review and edit the videos before they can be published. Tassoman's suggestion would be a good solution. 
That would also solve the problem that when a user leaves, all their content and channels are automatically gone.

**Comment by kimsible** [comment]
It's very sad to not include this feature in the 2021 roadmap, really it is asked by lots of instances... and since 2018, what kind of difficulties do you encounter @rigelk @Chocobozzz ? Federation ?

Collaborations channels is something very common now, unfortunately, the tool does not follow this, alternatively, we create a common account and we share a password, it's really bad for security and very inconvenient.

As I understood in discussions with Frama the last year, the 2021 roadmap ought to be focus on creators. New channels and accounts appearance follow this path. Why not this feature ? This may be one of the most important tool for creators.

Really, if this is just a question of time, Frama must rely on community for this, and if the community cannot achieve this, tips would be very welcomed from @Chocobozzz as you are the most indicated to bring precious and precious practices on the PeerTube source code.

At last there are other features links to this one asked the last year like the ownership channel transfer https://github.com/Chocobozzz/PeerTube/issues/1417 (some times duplicated) since 2018 too.

Hope my words will pay attention in the good way and encourage community to not be afraid to propose PRs.

**Comment by rigelk** [comment]
IMHO even without taking into account federation for this feature, we have intermediary roadblocks to clear. The RBAC system in use is one such roadblock that comes to mind: it limits access to ressources based on roles or being the owner, when we would need something more flexible.

A "quickfix" would be to have multiple owners, but arguably the AC system should be made more flexible first, otherwise we just implement a circumvention of our access control system for that specific use case.

**Comment by kimsible** [comment]
> IMHO even without taking into account federation for this feature, we have intermediary roadblocks to clear. The RBAC system in use is one such roadblock that comes to mind: it limits access to ressources based on roles or being the owner, when we would need something more flexible.
> 
> A "quickfix" would be to have multiple owners, but arguably the AC system should be made more flexible first, otherwise we just implement a circumvention of our access control system for that specific use case.

Thanks for this details @rigelk , maybe we can just have multiple authors (accounts) able to be mentioned displayed bellow the video next to the owner account for the case of collaborations and the main account stays the owner. A simple input like the keywords but for authors may be used. It would be a first step before changing the AC system, that I understand may take one more year (so let's be optimistic, maybe in the 2022 roadmap ?).

One last question, this PR https://github.com/Chocobozzz/PeerTube/issues/1417 about ownership channel transfer is also blocked for the same reason ?

**Comment by rigelk** [comment]
> maybe we can just have multiple authors (accounts) able to be mentioned displayed bellow the video next to the owner account for the case of collaborations and the main account publishing the video stays the owner. A simple input like the keywords but for authors may be used.

From what I understand, this is not related to access but just some added metadata to videos. So, not only can it be implemented before, but also after we come up with a implementation for the current issue.



> One last question, this PR #1417 about ownership channel transfer is also blocked for the same reason ?

What PR? As for the issue you link, it is not blocked by a not enough fine-grained access control system. IMHO it is "mostly" a question of having the proper UX since the channel actor stays the smae. We have already some example of UX with the video transfer, but that would need a serious factorization to have it for channels.

**Comment by kimsible** [comment]
> From what I understand, this is not related to access but just some added metadata to videos. So, not only can it be implemented before, but also after we come up with a implementation for the current issue.

A new issue can be created for this ? by the way I meant mentioning multiple accounts both for a video and for a channel.

> What PR? 

Sorry, I actually meant issue. 

> As for the issue you link, it is not blocked by a not enough fine-grained access control system. IMHO it is "mostly" a question of having the proper UX since the channel actor stays the smae. We have already some example of UX with the video transfer, but that would need a serious factorization to have it for channels.

Oh yes we discussed a little bit about it on Framacolibri the last year (forgot which subject).
Maybe  a stupid question but, channels transfer could proposed with a CLI tool, it would be firstly a request for admins, as I suppose implementing first this transfer on the server side could be easier than the client ?

**Comment by kontrollanten** [comment]
> IMHO even without taking into account federation for this feature, we have intermediary roadblocks to clear. The RBAC system in use is one such roadblock that comes to mind: it limits access to ressources based on roles or being the owner, when we would need something more flexible.

@rigelk have you thought of (or planned) any other access control system? Do you know any similar platform with an architecture that we would fit our needs? I'd be glad to help with the development, but it's not a typical feature that an one man army will solve in a PR.

**Comment by tutle** [comment]
Hello, it's a basic and essential fonction when some people want to share the admin of a channel (association, editing, commercial).
It's implemented in most of video plateform but seems not be possible in peertube.
I really need this, any plugin or possibility ?
Sincerely and thanks for this amazing job.

**Comment by pastoweb** [comment]
My 2 cents: this feature and a more fine permission system are indeed necessary.
Multiple users being able to manage or just publish on a channel would be essential for me.
Ready for financial contribution too

**Comment by KvaGram** [comment]
I just learned of Peertube. I am unsure if it can ever compete with youtube.
I share a youtube brand account with a few friends. We upload videos related to each other. It is convenient how every video ends up at the same youtube channel that way. Still it is annoying to have to swap account when uploading videos.
Peertube can improve on that by allowing for shared channels.

**Comment by pprw** [comment]
I just upvoted this feature. We also have the need of shared channels. I understand the complexity. Maybe a simple implementation could be to simply allow different users to upload into the same channel. Each user will be owner of his videos.

Aside of specific needs, IMHO it would be a good thing to promote collaborative creation. It would be another way to offer something different from Youtube (which is centered on individuals).

**Comment by biowilli** [comment]
I am thinking about to implement a channel permission system for Peertube. Here is a proposal:
As a user you could define permissions for each channel, such as who can upload, edit or delete videos.

We introduce two roles:

- channel-admin

permissions: is assigned automatically to the creator of the channel and s:he can add users with moderator permissions to each channel

- channel-moderator

permissions: edit or delete videos within that channel

As a channel-admin of a Peertube channel, I should be able to specify which users can upload videos, who can edit and delete videos to the channel. I should also be able to easily change the permissions for a channel at any point from the channel settings.
As a channel-moderator of a Peertube channel, I want to be able to upload, edit or delete videos within that channel.

We would need a table with a relation user - channel and add the role for each user. This results in the following extension of the database model:

![ER-Extension](https://user-images.githubusercontent.com/47250005/235621826-c4454771-f4fe-4a6a-af34-b5f81c5c82ce.png)

id: Unique ID for each authorisation
channelId: ID of the channel for which the authorisation is valid.
accountId: ID of the account to which the permission is granted.
permission: Type of permission

**Comment by pastoweb** [comment]
> I am thinking about to implement a channel permission system for Peertube. Here is a proposal: As a user you could define permissions for each channel, such as who can upload, edit or delete videos.
> 
> We introduce two roles:
> 
> * channel-admin
> 
> permissions: is assigned automatically to the creator of the channel and s:he can add users with moderator permissions to each channel
> 
> * channel-moderator
> 
> permissions: edit or delete videos within that channel
> 
> As a channel-admin of a Peertube channel, I should be able to specify which users can upload videos, who can edit and delete videos to the channel. I should also be able to easily change the permissions for a channel at any point from the channel settings. As a channel-moderator of a Peertube channel, I want to be able to upload, edit or delete videos within that channel.
> 
> We would need a table with a relation user - channel and add the role for each user. This results in the following extension of the database model:
> 
> ![ER-Extension](https://user-images.githubusercontent.com/47250005/235621826-c4454771-f4fe-4a6a-af34-b5f81c5c82ce.png)
> 
> id: Unique ID for each authorisation channelId: ID of the channel for which the authorisation is valid. accountId: ID of the account to which the permission is granted. permission: Type of permission

Sounds good, hope chocobozz will consider adding it to the main platform. Are you going to release it as a plugin?

**Comment by biowilli** [comment]
> > I am thinking about to implement a channel permission system for Peertube. Here is a proposal: As a user you could define permissions for each channel, such as who can upload, edit or delete videos.
> > We introduce two roles:
> > 
> > * channel-admin
> > 
> > permissions: is assigned automatically to the creator of the channel and s:he can add users with moderator permissions to each channel
> > 
> > * channel-moderator
> > 
> > permissions: edit or delete videos within that channel
> > As a channel-admin of a Peertube channel, I should be able to specify which users can upload videos, who can edit and delete videos to the channel. I should also be able to easily change the permissions for a channel at any point from the channel settings. As a channel-moderator of a Peertube channel, I want to be able to upload, edit or delete videos within that channel.
> > We would need a table with a relation user - channel and add the role for each user. This results in the following extension of the database model:
> > ![ER-Extension](https://user-images.githubusercontent.com/47250005/235621826-c4454771-f4fe-4a6a-af34-b5f81c5c82ce.png)
> > id: Unique ID for each authorisation channelId: ID of the channel for which the authorisation is valid. accountId: ID of the account to which the permission is granted. permission: Type of permission
> 
> Sounds good, hope chocobozz will consider adding it to the main platform. Are you going to release it as a plugin?

I would like to implement it to the main platform. I think it isn't possible as a plugin :)

**Comment by kontrollanten** [comment]
> I will implement a PoC in peertube core. I think it isn't possible as a plugin :) @pastoweb

How is it going, @biowilli ? This feature would be much appreciated!

Regarding your proposal of permission model, I think channel-admin and channel-moderator should be able to remove comments as well? I think that a channel owner can do this today. There’s also description, banner, avatar and playlists that should have shared permissions as well.

**Comment by biowilli** [comment]
> 
> How is it going, @biowilli ? This feature would be much appreciated!
> 

Progress on my end has unfortunately been minimal :(

**Comment by kpideas** [comment]
+1 this would be an amazing feature.

For a start, it would be great even if we can just allow multiple users posting videos on Shared Channels. More advanced Channel management can come later. 

I was checking the database schema. "videoChannel table has accountId field which is the primary user who created that channel. Maybe we could introduce a new string/varchar field to capture comma separated account ids who has the rights to post videos on that channel? 

So that on the Publish screen, the channel drop down can have self created channels and the channels where the current user has been granted permissions? 

and On channel page we can have list of secondary users like labels.

Of course creating a child table and doing proper relationships is better way, was just thinking a quick win. Even with child table solutions proposed above, we could probably start with allowing multiple users to post videos on shared channels and more advanced channel admin features can come later?

**Comment by stevespaw** [comment]
> > How is it going, [@biowilli](https://github.com/biowilli) ? This feature would be much appreciated!
> 
> Progress on my end has unfortunately been minimal :(

We have clients that could really use this. Several departments that all need to contribute to the same channel. And need to manage those users in the system as they come and go.

**Comment by Chocobozzz** [comment]
Framasoft may have funding so we can work on this feature this year. It would be a lighter feature where a channel owner would be able to add moderators to it, so that they could also upload and manage channel content. We'll keep you informed :)

**Comment by Chocobozzz** [comment]
Framasoft may have funding so we can work on this feature this year. It would be a lighter feature where a channel owner would be able to add moderators to it, so that they could also upload and manage channel content. We'll keep you informed :)

**Comment by T1stHaunt3R** [comment]
Hi @Chocobozzz, is there any ETA for this feature? Or is there a place we can see the roadmap?

**Comment by Chocobozzz** [comment]
Roadmap is available on https://joinpeertube.org/news/roadmap-2025

This specific feature will be released in PeerTube v8, planned for December.

**Comment by Chocobozzz** [comment]
Implemented in RC of v8 🎉  https://github.com/Chocobozzz/PeerTube/releases/tag/v8.0.0-rc.1
